### ISSUE TITLE
Lintr to enforce :: for packages

### ISSUE THREAD

**Issue by hhoeflin** [jimhester/lintr#459]
Hi,

I was wondering what you/the community thinks about a linter enforcing use of '::'. In python, code is often much easier to read as it is simpler to figure out where a function is coming from.

So the idea would be to have a linter that
- reports any function from a package that is used without a '::'
  - this could be disabled by a comment along the lines of: 'from package import blah', i.e. similar to python, but based on a comment as it is not a language feature
  - for 'base' packages this could be relaxed or switched off, or generally switched off for a defined set of packages
- functions used in a script that are from another script (e.g. inside a package) could also be declared in comments

Intent would be to increase general readability and encourage defensive programming (with the '::')

Thoughts?

**Comment by russHyde** [comment]
Crossref: #114

**Comment by MichaelChirico** [comment]
I would still accept a PR creating such a linter, but FWIW I am increasingly against this idea.

The core of the problem is that it encourages run-time dependencies where load-time dependencies are more appropriate.

Authors insisting on `::` would be well-advised to _also_ be sure to include relevant packages in their package NAMESPACE with `import()` or `importFrom()` directives to force any `.onLoad()` hooks to run at load time.

But then using `::` is kind of redundant since the NAMESPACE can be referenced as needed to help resolving symbols.

We recently did a sweep of {lintr}'s own codebase to remove `::` usage for many highly-used calls: #2031 and #2079. A compromise was struck there where foreign calls only used a few times are kept in `::` usage for emphasis/clarity.

**Comment by salim-b** [comment]
> Authors insisting on `::` would be well-advised to _also_ be sure to include relevant packages in their package NAMESPACE with `import()` or `importFrom()` directives to force any `.onLoad()` hooks to run at load time.

Side question out of sincere curiosity: Why exactly would that be good advice? Without `import()` or `importFrom()` directives in the package NAMESPACE, the `.onLoad()` hooks are simply run at run time (when for the first time a function from a package is invoked in the current R session). I did never experience any issue with this so far...

**Comment by salim-b** [comment]
Thank you @TimTaylor for the illustrating example! True about the earliness of the warning. But I don't think that's an actual advantage. If a "rarely called function" F1 from pkg A is broken because it calls a function F2 from pkg B that doesn't exist anymore, it will only become an issue for users once they actually try to use F1. So IMHO it's totally fine if they're only warned then and not earlier. Also in both cases F1 stays broken until it's actually adapted to not call F2 anymore, so a warning might also cause unwanted noise in issue trackers etc.

> This is less likely with CRAN packages

Yep, cause `R CMD check` will tell you about the inexisting F2 from package B.

**Comment by MichaelChirico** [comment]
> Without `import()` or `importFrom()` directives in the package NAMESPACE, the `.onLoad()` hooks are simply run at run time

Right, but these might set some state that's needed/beneficial earlier.

> Would this be possible to lint for?

I think it's hard to distinguish between Suggests usage of `::` and Imports usage statically.

It's also hard-ish to detect when `::` _must_ be used because of upstream conflicts, e.g. if _both_ `stats::filter()` and `dplyr::filter()` are needed by the same package, only one can possibly be in NAMESPACE --> one will always need to be qualified, or at very least have something like `stats_filter = stats::filter` somewhere. E.g. here's such a package on CRAN that uses both functions:

https://github.com/cran/bayesplot/blob/557a0e687f87985af861c080054c6afcdb8dcc3b/R/ppc-loo.R

**Comment by salim-b** [comment]
> Right, but these might set some state that's needed/beneficial earlier.

I'd like to know more about this. I can't really imagine a situation where this is actually the case (it would mean package B is setting state for *others* besides itself, which seems like a terrible design to me).

**Comment by AshesITR** [comment]
One example could be `{rJava}` which has expensive `.onLoad()`. This would make the first function call unexpectedly slow if the caller doesn't use rJava elsewhere.

Also, if namespace loading fails for some reason, the earlier this happens, the less compute time is wasted before the user is alerted to this problem.

**Comment by MichaelChirico** [comment]
More thoughts here: with `testthat::local_mocked_binding()` replacing the old {mockery} (etc) ways of test mocking, using `::` in packages is even further discouraged:

https://testthat.r-lib.org/reference/local_mocked_bindings.html#namespaced-calls

It's only possible to mock out names in the NAMESPACE, not as `::`-qualified calls.
